File: Haskell/ListsAndArrays/Find/ArrayElementEqIndex/src/ArrayElementEqIndex.hs
module ArrayElementEqIndex (arrayElementEqIndex) where

import Data.Array (Array, (!), bounds)

-- Given a sorted array (smallest to largest) of n distinct integers
-- (positive, negative, or zero), return True if there exists an
--  index i such that the element at that index is equal to the index.

-- INPUT: A sorted array of n distinct integers, arr.
-- OUTPUT: True if there is i s.t. arr ! i == i. False otherwise.
-- ASSUMES: 1-indexing in the array.
arrayElementEqIndex :: Array Integer Integer -> Bool
arrayElementEqIndex arr =
    -- Start the search with the full range of the array.
    go (bounds arr)
    where  
        go (il, iu)
                -- Base case (n=1): Check if the one element meets the condition.
                | il == iu = arr ! il == il
                -- Base case (n=2): Check if either element meets the condition.
                | abs (il - iu) < 2 = 
                    arr ! il == il || arr ! iu == iu
                -- Inductive step:
                | otherwise = 
                    -- Choose a point in the middle of the search window.
                    let midPoint = (il + iu) `div` 2 in
                    -- Compare A[i] to i:
                    case compare (arr ! midPoint) midPoint of
                        -- If they equal, the condition is met.
                        EQ -> True
                        -- If A[i] > i, then the condition cannot be met,
                        --  since the integers are sorted and distinct.
                        GT -> False
                        -- If A[i] < i, the condition could possibly be met in
                        -- the right half of the search window.
                        LT -> go (midPoint, iu)

File: Java/src/ListsAndArrays/Find/ArrayElementEqIndex.java
package ListsAndArrays.Find;

public class ArrayElementEqIndex {

    /** Given a sorted array (smallest to largest) of n distinct integers
        (positive, negative, or zero), return True if there exists an
        index i such that the element at that index is equal to the index.

        INPUT: A sorted array of n distinct integers, arr.
        OUTPUT: True if there is i s.t. arr ! i == i. False otherwise.
        ASSUMES: 0-indexing in the array. */
    static Boolean arrayElementEqIndex(int[] arr) {

        int n = arr.length;
        int windowStart = 0;
        int windowEnd = n - 1;
        int midPoint;

        switch (n) {

            case 1:
                return arr[0] == 0;

            case 2:
                return arr[0] == 0 || arr[1] == 1;

            default:

                while (windowEnd - windowStart > 1) {

                    midPoint = (windowStart + windowEnd) / 2;

                    if (arr[midPoint] == midPoint) {
                        return true;
                    }
                    else if (arr[midPoint] > midPoint) {
                        return false;
                    }
                    else {
                        windowStart = midPoint;
                    }
                }
                return arr[windowStart] == windowStart || arr[windowEnd] == windowEnd;
        }
    }
}
